File: lyrical/views/api_summarise_chat_history.py
from django.http import JsonResponse
from django.contrib.auth.decorators import login_required
from .. import models
from ..services.utils.summarise import ChatSummarisationService
import json
import logging
from ..logging_config import get_logger


logger = get_logger('apis')


@login_required
def api_summarise_chat_history(request):
    """
    Summarise chat history for a song and message type.
    
    This endpoint will:
    1. Check if summarisation is needed based on token count
    2. Generate a summary using the user's summarisation LLM model
    3. Deactivate original messages and insert summary
    4. Update the song's needs_summarisation flag
    
    Args:
        request: The HTTP request object containing POST data with song_id and message_type
    
    Returns:
        JsonResponse: Success/error response with summarisation details
    """
    # Validate request method
    if request.method != 'POST':
        return JsonResponse({"error": "Method not allowed"}, status=405)

    try:
        # Parse the request body
        request_data = json.loads(request.body.decode('utf-8'))

        logger.debug(f"Chat summarisation request data: {json.dumps(request_data, indent=2)}")

        song_id = request_data.get("song_id")
        message_type = request_data.get("message_type")
        force_summarise = request_data.get("force_summarise", False)

        # Validate required parameters
        if not song_id:
            return JsonResponse({"error": "Song ID must be provided"}, status=400)
        
        if not message_type:
            return JsonResponse({"error": "Message type must be provided"}, status=400)
        
        if message_type not in ['style', 'lyrics']:
            return JsonResponse({"error": "Message type must be 'style' or 'lyrics'"}, status=400)

        # Verify song exists and user has permission
        try:
            song = models.Song.objects.get(id=song_id, user=request.user)
        except models.Song.DoesNotExist:
            return JsonResponse({
                "error": f"Song with ID {song_id} not found or you don't have permission to access it"
            }, status=404)

        # Check if summarisation is actually needed (unless forced)
        if not force_summarise:
            needs_summary = ChatSummarisationService.check_needs_summarisation(
                song_id, message_type, request.user
            )
            if not needs_summary:
                return JsonResponse({
                    "status": "not_needed",
                    "message": f"Conversation for {message_type} does not need summarisation yet",
                    "song_id": song_id,
                    "message_type": message_type
                }, status=200)

        # Get conversation stats before summarisation
        from ..services.message_history_service import MessageHistoryService
        stats_before = MessageHistoryService.get_conversation_stats(song_id, message_type, request.user)

        logger.info(f"Starting chat summarisation for user {request.user.username}, song {song_id}, type '{message_type}'")
        logger.debug(f"Conversation stats before summarisation: {stats_before}")

        # Perform the summarisation
        success = ChatSummarisationService.summarise_conversation(
            song_id, message_type, request.user
        )

        if not success:
            logger.error(f"Summarisation failed for user {request.user.username}, song {song_id}, type '{message_type}'")
            return JsonResponse({
                "error": "Failed to summarise conversation. Please try again later."
            }, status=500)

        # Get conversation stats after summarisation
        stats_after = MessageHistoryService.get_conversation_stats(song_id, message_type, request.user)

        logger.info(f"Chat summarisation completed for user {request.user.username}, song {song_id}, type '{message_type}'")
        logger.debug(f"Conversation stats after summarisation: {stats_after}")

        # Check if song still needs summarisation (for other conversation types)
        final_needs_summary = ChatSummarisationService.update_song_summarisation_flag(
            song_id, request.user
        )

        response_data = {
            "status": "success",
            "message": f"Successfully summarised {message_type} conversation",
            "song_id": song_id,
            "message_type": message_type,
            "stats_before": stats_before,
            "stats_after": stats_after,
            "messages_summarised": stats_before.get('active_total_messages', 0),
            "song_still_needs_summarisation": song.needs_summarisation
        }

        logger.info(f"Summarisation response: {json.dumps(response_data, indent=2)}")
        return JsonResponse(response_data, status=200)

    except json.JSONDecodeError:
        logger.error(f"Invalid JSON in summarisation request from user {request.user.username}")
        return JsonResponse({"error": "Invalid JSON data provided"}, status=400)
    
    except Exception as e:
        logger.error(f"Unexpected error during chat summarisation for user {request.user.username}: {str(e)}")
        import traceback
        logger.error(f"Traceback: {traceback.format_exc()}")
        return JsonResponse({
            "error": "An unexpected error occurred during summarisation. Please try again later."
        }, status=500)


@login_required
def api_check_summarisation_status(request):
    """
    Check if any conversations for a song need summarisation.
    
    Args:
        request: The HTTP request object with song_id parameter
    
    Returns:
        JsonResponse: Status information about summarisation needs
    """
    if request.method != 'GET':
        return JsonResponse({"error": "Method not allowed"}, status=405)

    try:
        song_id = request.GET.get("song_id")
        
        if not song_id:
            return JsonResponse({"error": "Song ID must be provided"}, status=400)

        # Verify song exists and user has permission
        try:
            song = models.Song.objects.get(id=song_id, user=request.user)
        except models.Song.DoesNotExist:
            return JsonResponse({
                "error": f"Song with ID {song_id} not found or you don't have permission to access it"
            }, status=404)

        # Check summarisation status for both conversation types
        style_needs_summary = ChatSummarisationService.check_needs_summarisation(
            song_id, 'style', request.user
        )
        lyrics_needs_summary = ChatSummarisationService.check_needs_summarisation(
            song_id, 'lyrics', request.user
        )

        # Get conversation stats
        from ..services.message_history_service import MessageHistoryService
        style_stats = MessageHistoryService.get_conversation_stats(song_id, 'style', request.user)
        lyrics_stats = MessageHistoryService.get_conversation_stats(song_id, 'lyrics', request.user)

        response_data = {
            "song_id": song_id,
            "song_needs_summarisation": song.needs_summarisation,
            "style": {
                "needs_summarisation": style_needs_summary,
                "stats": style_stats
            },
            "lyrics": {
                "needs_summarisation": lyrics_needs_summary,
                "stats": lyrics_stats
            },
            "any_needs_summarisation": style_needs_summary or lyrics_needs_summary
        }

        return JsonResponse(response_data, status=200)

    except Exception as e:
        logger.error(f"Error checking summarisation status for user {request.user.username}: {str(e)}")
        return JsonResponse({
            "error": "Failed to check summarisation status. Please try again later."
        }, status=500)

File: lyrical/static/lyrical/script/api_summarise_chat_history.js
/**
 * API functions for chat history summarisation.
 * Provides methods to check summarisation status and trigger summarisation.
 */

/**
 * Check if any conversations for a song need summarisation
 * @param {string|number} songId - The ID of the song to check
 * @returns {Promise<Object>} Promise that resolves to summarisation status information
 */
export function apiCheckSummarisationStatus(songId) {
    const url = new URL('/api_check_summarisation_status', window.location.origin);
    url.searchParams.append('song_id', songId);

    console.log('Checking summarisation status for song:', songId);

    return fetch(url, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
    })
    .then(response => {
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return response.json();
    })
    .then(data => {
        console.log('Summarisation status check result:', data);
        return data;
    })
    .catch(error => {
        console.error('Error checking summarisation status:', error);
        throw error;
    });
}

/**
 * Summarise chat history for a specific conversation type
 * @param {string|number} songId - The ID of the song
 * @param {string} messageType - The type of conversation ('style' or 'lyrics')
 * @param {boolean} [forceSummarise=false] - Whether to force summarisation even if not needed
 * @returns {Promise<Object>} Promise that resolves to summarisation result
 */
export function apiSummariseChatHistory(songId, messageType, forceSummarise = false) {
    // Validate message type
    if (!['style', 'lyrics'].includes(messageType)) {
        return Promise.reject(new Error('Message type must be "style" or "lyrics"'));
    }

    // Get CSRF token
    const csrfToken = document.querySelector('[name=csrfmiddlewaretoken]').value;

    // Build request body
    const requestBody = {
        song_id: songId,
        message_type: messageType,
        force_summarise: forceSummarise
    };

    console.log('Summarisation request:', JSON.stringify(requestBody, null, 2));

    return fetch('/api_summarise_chat_history', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'X-CSRFToken': csrfToken,
        },
        body: JSON.stringify(requestBody)
    })
    .then(response => {
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return response.json();
    })
    .then(data => {
        console.log('Summarisation result:', data);
        
        if (data.status === 'success') {
            console.log(`Successfully summarised ${data.messages_summarised} messages for ${messageType} conversation`);
            return data;
        } else if (data.status === 'not_needed') {
            console.log(`Summarisation not needed for ${messageType} conversation`);
            return data;
        } else {
            throw new Error(data.error || 'Unknown error occurred during summarisation');
        }
    })
    .catch(error => {
        console.error('Error summarising chat history:', error);
        throw error;
    });
}

/**
 * Summarise all conversations for a song that need summarisation
 * @param {string|number} songId - The ID of the song
 * @param {boolean} [forceSummarise=false] - Whether to force summarisation even if not needed
 * @returns {Promise<Object>} Promise that resolves to combined summarisation results
 */
export function apiSummariseAllConversations(songId, forceSummarise = false) {
    console.log('Starting summarisation for all conversations for song:', songId);

    // First check what needs summarisation
    return apiCheckSummarisationStatus(songId)
        .then(status => {
            const promises = [];
            const results = {
                song_id: songId,
                style: null,
                lyrics: null,
                any_summarised: false
            };

            // Summarise style conversation if needed or forced
            if (forceSummarise || status.style.needs_summarisation) {
                console.log('Queueing style conversation for summarisation');
                promises.push(
                    apiSummariseChatHistory(songId, 'style', forceSummarise)
                        .then(result => {
                            results.style = result;
                            if (result.status === 'success') {
                                results.any_summarised = true;
                            }
                        })
                        .catch(error => {
                            results.style = { status: 'error', error: error.message };
                        })
                );
            } else {
                console.log('Style conversation does not need summarisation');
                results.style = { status: 'not_needed', message: 'Style conversation does not need summarisation' };
            }

            // Summarise lyrics conversation if needed or forced
            if (forceSummarise || status.lyrics.needs_summarisation) {
                console.log('Queueing lyrics conversation for summarisation');
                promises.push(
                    apiSummariseChatHistory(songId, 'lyrics', forceSummarise)
                        .then(result => {
                            results.lyrics = result;
                            if (result.status === 'success') {
                                results.any_summarised = true;
                            }
                        })
                        .catch(error => {
                            results.lyrics = { status: 'error', error: error.message };
                        })
                );
            } else {
                console.log('Lyrics conversation does not need summarisation');
                results.lyrics = { status: 'not_needed', message: 'Lyrics conversation does not need summarisation' };
            }

            // Wait for all summarisations to complete
            return Promise.all(promises).then(() => results);
        })
        .then(results => {
            console.log('All conversation summarisation completed:', results);
            return results;
        });
}

/**
 * Show a user notification about summarisation status.
 * This function can be used to inform users when summarisation is needed or completed.
 * @param {Object} statusData - Status data from apiCheckSummarisationStatus
 * @param {Function} notificationFunction - Function to display notifications (e.g., showErrorToast)
 */
export function showSummarisationNotification(statusData, notificationFunction) {
    if (!statusData.any_needs_summarisation) {
        return; // No notification needed
    }

    const conversationsNeedingSummary = [];
    if (statusData.style.needs_summarisation) {
        conversationsNeedingSummary.push('style');
    }
    if (statusData.lyrics.needs_summarisation) {
        conversationsNeedingSummary.push('lyrics');
    }

    const message = `Your ${conversationsNeedingSummary.join(' and ')} conversation${conversationsNeedingSummary.length > 1 ? 's' : ''} ${conversationsNeedingSummary.length > 1 ? 'are' : 'is'} getting long. Consider summarising to improve performance.`;
    
    if (notificationFunction) {
        notificationFunction(message);
    } else {
        console.warn('Summarisation needed:', message);
    }
}